devices/src/virtio/vsock/csm/connection.rs

```diff
@@ -0,0 +1,632 @@
+    /// was ready to be read from the host stream.
+    ///
+    /// Returns:
+    /// - `Ok(())`: the packet has been successfully filled in and is ready for delivery;
+    /// - `Err(VsockError::NoData)`: there was no data available with which to fill in the packet;
+    /// - `Err(VsockError::PktBufMissing)`: the packet would've been filled in with data, but it is
+    ///   missing the data buffer.
+    ///
+    fn recv_pkt(&mut self, pkt: &mut VsockPacket) -> VsockResult<()> {
+        // Perform some generic initialization that is the same for any packet operation (e.g.
+        // source, destination, credit, etc).
+        self.init_pkt(pkt);
+
+        // If forcefull termination is pending, there's no point in checking for anything else.
+        // It's dead, Jim.
+        if self.pending_rx.remove(PendingRx::Rst) {
+            pkt.set_op(uapi::VSOCK_OP_RST);
+            return Ok(());
+        }
+
+        // Next up: if we're due a connection confirmation, that's all we need to know to fill in
+        // this packet.
+        if self.pending_rx.remove(PendingRx::Response) {
+            self.state = ConnState::Established;
+            pkt.set_op(uapi::VSOCK_OP_RESPONSE);
+            return Ok(());
+        }
+
+        // Same thing goes for locally-initiated connections, that need to yield a connection request.
+        if self.pending_rx.remove(PendingRx::Request) {
+            self.expiry =
+                Some(Instant::now() + Duration::from_millis(defs::CONN_REQUEST_TIMEOUT_MS));
+            pkt.set_op(uapi::VSOCK_OP_REQUEST);
+            return Ok(());
+        }
+
+        // A credit update is basically a no-op, so we should only waste a perfectly fine RX buffer
+        // on it if we really have nothing else to say.
+        if self.pending_rx.remove(PendingRx::CreditUpdate) && !self.has_pending_rx() {
+            pkt.set_op(uapi::VSOCK_OP_CREDIT_UPDATE);
+            self.last_fwd_cnt_to_peer = self.fwd_cnt;
+            return Ok(());
+        }
+
+        // Alright, if we got to here, we need to cough up a data packet. We've already checked for
+        // all other pending RX indications.
+        self.pending_rx.remove(PendingRx::Rw);
+
+        match self.state {
+            // A data packet is only valid for established connections, and connections for which
+            // our peer has initiated a graceful shutdown, but can still receive data.
+            ConnState::Established | ConnState::PeerClosed(false, _) => (),
+            _ => {
+                // Any other connection state is invalid at this point, and we need to kill it with
+                // fire.
+                pkt.set_op(uapi::VSOCK_OP_RST);
+                return Ok(());
+            }
+        }
+
+        // Oh wait, before we start bringing in the big data, can our peer handle receiving so much
+        // bytey goodness?
+        if self.need_credit_update_from_peer() {
```

Comment:
Didn't not think this through very well :D, but just wondering, is it possible to generate a lot of ops/traffic because we keep calling `recv_pkt` because there's pending data, but we don't know if the goodness can be handled, so we also keep sending these requests, until one finally comes back? So we no longer attempt to send anything if we've sent a credit request before receiving the associated response? Does it make sense to worry about this? :-s